### PR TITLE
Drv flow

### LINKED ISSUE #42: Enhance drv_flow
- [x] Document conection:
https://github.com/WattRex/Battery-Cycler-Drivers/blob/7ff7b02d86ce8817c3ee3499c85a47b8b1848a93/code/drv_flow/README.md?plain=1#L2
- [x] Enhance firmware arduino. Add device number constant and firmware number version.
- [x] Implement _read-device function on driver
- [x] Adapt arch to allow conection with scpi-sniffer
- [x] Integrate SCPI-sniffer

### PR DESCRIPTION
**Pull request by luroche**
Working drv flow

### CODE REVIEW

File: code/drv_flow/resources/flowmeter_arduino.ino [hunk 1]
```diff
@@ -1,78 +1,62 @@
-// Digital pin used for negative flowmeter
-#define pin_neg 2
-
+// Digital pin used for auxiliar flowmeter
+#define pin_aux 2
 #define FACTOR_MILIS 1000
 #define SACLAR_US_2_FREQ 500000 // Factor to convert us to Hz -> 1.000.000/ (2 * semiperiod)
 #define FLOW_MAX 10 // 10 L/min 
 #define FREQ_MAX 235 // Freq for 10L/min
+#define DEVICE_NUMBER "001"
```
Comment (mariuscrsn) [inline code/drv_flow/resources/flowmeter_arduino.ino]: Change to int

File: code/drv_flow/resources/flowmeter_arduino.ino [hunk 2]
```diff
@@ -1,78 +1,62 @@
...
-#define pin_neg 2
-
+// Digital pin used for auxiliar flowmeter
+#define pin_aux 2
 #define FACTOR_MILIS 1000
 #define SACLAR_US_2_FREQ 500000 // Factor to convert us to Hz -> 1.000.000/ (2 * semiperiod)
 #define FLOW_MAX 10 // 10 L/min 
 #define FREQ_MAX 235 // Freq for 10L/min
+#define DEVICE_NUMBER "001"
+#define FIRMWARE_VERSION "1"
```
Comment (mariuscrsn) [inline code/drv_flow/resources/flowmeter_arduino.ino]: change to int

File: code/drv_flow/resources/flowmeter_arduino.ino [hunk 3]
```diff
@@ -1,78 +1,62 @@
...
 void setup() {
 
-  Serial.begin(9600);
+  Serial.begin(BAUDRATE);
 
   while (!Serial) {
     Serial.println("Init flow meter");
   }
-
-  pinMode(pin_neg, INPUT);
```
Comment (mariuscrsn) [inline code/drv_flow/resources/flowmeter_arduino.ino:43]: Not remove it

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 4]
```diff
@@ -4,78 +4,207 @@
...
+path.append(os.getcwd())
+from system_logger_tool import SysLogLoggerC, sys_log_logger_get_module_logger # pylint: disable=wrong-import-position
+if __name__ == '__main__':
+    cycler_logger = SysLogLoggerC(file_log_levels='./log_config.yaml')
+log = sys_log_logger_get_module_logger(__name__)
+from system_shared_tool import SysShdIpcChanC, SysShdNodeStatusE # pylint: disable=wrong-import-position
 
 #######################          PROJECT IMPORTS         #######################
-
+from scpi_sniffer import *
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Import only commands

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 5]
```diff
@@ -4,78 +4,207 @@
...
-class _CONSTANTS:
-    SERIAL_PORT = "/dev/ttyACM1"
-    MILI_UNITS = 1000
-    MAX_ATTEMPTS = 5
+MAX_WAIT_TIME = 5
+TIME_BETWEEN_ATTEMPTS = 0.1
+MAX_MSG = 100
+MAX_MESSAGE_SIZE = 300
 
 class ScpiCmds(Enum):
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Cast it to private Enum: `_ScpiCmds`

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 6]
```diff
@@ -4,78 +4,207 @@
...
+        Raises:
+            - None.
+        '''
+        del_msg = DrvScpiCmdDataC(data_type = DrvScpiCmdTypeE.DEL_DEV,
+                                  port = self.__port) # pylint: disable=no-member
+        self.__tx_chan.send_data(del_msg)
+        self.__rx_chan.terminate()
+
+
+    def __signal_handler(self, sig, frame) -> None: # pylint: disable=unused-argument
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: remove it

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 3 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino:43]: Done.

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 2 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino]: remains as string

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 1 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino]: remains as string

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 3 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino:43]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 7]
```diff
@@ -4,78 +4,207 @@
...
+        add_msg = DrvScpiCmdDataC(data_type = DrvScpiCmdTypeE.ADD_DEV,
+                                  port = config.port,
+                                  payload = config,
+                                  rx_chan_name = rx_chan_name)
+        self.__rx_chan.delete_until_last()
+        self.__tx_chan.send_data(add_msg)
+        self.__last_meas: DrvFlowDataC = DrvFlowDataC(flow_main = 0, flow_aux = 0)
+        self.__wait_4_response = False
+        self.__read_device_properties()
+        signal(SIGINT, self.__signal_handler)
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: remove it

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 8]
```diff
@@ -4,78 +4,207 @@
...
+
+        # Wait until receive the message
+        time_init = time()
+        while (time() - time_init) < MAX_WAIT_TIME:
+            sleep(TIME_BETWEEN_ATTEMPTS)
+            if not self.__rx_chan.is_empty():
+                command_rec : DrvScpiCmdDataC = self.__rx_chan.receive_data()
+                log.critical(f"Command received: {command_rec.payload}")
+
+                # failed = command_rec.status == SysShdNodeStatusE.COMM_ERROR
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: remove it

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 9]
```diff
@@ -4,78 +4,207 @@
...
+        if not self.__wait_4_response:
+            msg = DrvScpiCmdDataC(data_type = DrvScpiCmdTypeE.WRITE_READ,
+                                port = self.__port,
+                                payload = ScpiCmds.GET_MEAS.value + '\n')
+            log.info(f"Message to send: {msg.payload}")
+            self.__tx_chan.send_data(msg)
+            self.__wait_4_response = True
+        else:
+            if not self.__rx_chan.is_empty():
+                data : DrvScpiCmdDataC = self.__rx_chan.receive_data()
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Receive with an timeout

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 10]
```diff
@@ -4,78 +4,207 @@
...
+                msg_received = data.payload[0]
+                log.critical(f"Message received: {msg_received}")
+                msg_received = msg_received.split(':')
+                self.__wait_4_response = False
+                if len(msg_received) > 0 and ('ERROR' not in data.payload) and \
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Update status in last_meas y receive an ERROR

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 11]
```diff
@@ -4,78 +4,207 @@
...
+                                payload = ScpiCmds.GET_MEAS.value + '\n')
+            log.info(f"Message to send: {msg.payload}")
+            self.__tx_chan.send_data(msg)
+            self.__wait_4_response = True
+        else:
+            if not self.__rx_chan.is_empty():
+                data : DrvScpiCmdDataC = self.__rx_chan.receive_data()
+                msg_received = data.payload[0]
+                log.critical(f"Message received: {msg_received}")
+                msg_received = msg_received.split(':')
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Split message after checking length of msg_received

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 4 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done,

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 5 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 6 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 7 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 8 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 9 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 11 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 10 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 2 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py [hunk 12]
```diff
@@ -160,24 +154,24 @@ def get_meas(self) -> DrvFlowDataC:
...
-                self.__wait_4_response = False
-                if len(msg_received) > 0 and ('ERROR' not in data.payload) and \
-                    ('IDN' not in data.payload):
-                    # f_main = int(msg_received[msg_received.index('DATA')+1])
-                    # f_aux = int(msg_received[msg_received.index('DATA')+2])
-                    f_main = 0
-                    f_aux = 0
-                    self.__last_meas = DrvFlowDataC(flow_main = f_main, flow_aux = f_aux)
+            time_init = time()
+            while (time() - time_init) < _MAX_WAIT_TIME:
```
Comment (mariuscrsn) [inline code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Remove while

Comment (mariuscrsn) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: and add timeout to receive_data

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 5 above)
Comment (mariuscrsn) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py:36]: Change enum name

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/drv_flow/resources/flowmeter_arduino.ino (hunk 1 above)
Comment (luroche) [reply code/drv_flow/resources/flowmeter_arduino.ino]: Done.

File: code/drv_flow/src/wattrex_driver_flow/drv_flow.py (hunk 5 above)
Comment (luroche) [reply code/drv_flow/src/wattrex_driver_flow/drv_flow.py]: Done.

**Review by mariuscrsn — APPROVED** [review]

**Pull request merged**